resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -86,21 +83,9 @@ private[yarn] class YarnAllocator(
 
   private val numExecutorsStarting = new AtomicInteger(0)
 
-  /**
-   * Used to generate a unique ID per executor
-   *
-   * Init `executorIdCounter`. when AM restart, `executorIdCounter` will reset to 0. Then
-   * the id of new executor will start from 1, this will conflict with the executor has
-   * already created before. So, we should initialize the `executorIdCounter` by getting
-   * the max executorId from driver.
-   *
-   * And this situation of executorId conflict is just in yarn client mode, so this is an issue
-   * in yarn client mode. For more details, can check in jira.
-   *
-   * @see SPARK-12864
-   */
-  private var executorIdCounter: Int =
-    driverRef.askSync[Int](RetrieveLastAllocatedExecutorId)
```

Comment:
Did `RetrieveLastAllocatedExecutor` always return `0`  if no executors had been launched? Or was there a `-1 ` case so that the IDs would start at 0 after being incremented in `runAllocatedContainers`?

Reply:
The reverted commit introduced a field `currentExecutorIdCounter = 0` in `CoarseGrainedSchedulerBackend` which is extended by `YarnSchedulerBackend`. `YarnSchedulerBackend` in turn, returns this number when receives `RetrieveLastAllocatedExecutorId` message. Please check out [this comment](https://github.com/mesosphere/spark/pull/37#discussion_r226031446) regarding the logic behind the reverted commit.